File: 694.number-of-distinct-islands.js
/**
 * 
Example 1:


Input: grid = [[1,1,0,0,0],[1,1,0,0,0],[0,0,0,1,1],[0,0,0,1,1]]
Output: 1
 */

/**
 * @param {number[][]} grid
 * @return {number}
 */
var numDistinctIslands = function (grid) {
  const m = grid.length;
  const n = grid[0].length;

  const dfs = (x, y) => {
    if (!grid[x]?.[y]) return [];
    grid[x][y] = null;
    const r1 = dfs(x + 1, y);
    const r2 = dfs(x - 1, y);
    const r3 = dfs(x, y + 1);
    const r4 = dfs(x, y - 1);
    return [{ x, y }, ...r1, ...r2, ...r3, ...r4];
  };

  const set = new Set();
  for (let i = 0; i < m; ++i) {
    for (let j = 0; j < n; ++j) {
      if (grid[i][j]) {
        const island = dfs(i, j);
        const key = island.map(({ x, y }) => `${x - i},${y - j}`).join(",");
        set.add(key);
      }
    }
  }
  return set.size;
};

File: solutions/694.number-of-distinct-islands.rb
# @param {Integer[][]} grid
# @return {Integer}
def num_distinct_islands(grid)
  m = grid.size
  n = grid[0].size
  island_set = Set.new
  dfs = lambda do |x, y|
    return [] if [x < 0, y < 0, x == m, y == n].any? || grid[x][y] == 0

    grid[x][y] = 0
    down = dfs.call(x - 1, y)
    up = dfs.call(x + 1, y)
    left = dfs.call(x, y - 1)
    right = dfs.call(x, y + 1)
    [[x, y]] + down + up + left + right
  end

  (0...m).each do |i|
    (0...n).each do |j|
      next if (island = dfs.call(i, j)).empty?

      island_set << island.map { |x, y| [x - i, y - j] }
    end
  end

  island_set.size
end
